plutarch-context-builder/src/Plutarch/Context/Transformer.hs

```diff
@@ -0,0 +1,375 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Plutarch.Context.Transformer (
+  -- * Data type
+  Transformer,
+  -- * Introduction
+  -- ** Generic
+  liftDatumEndo,
+  liftDatumKleisli,
+  liftDatumCheck,
+  liftRedeemerEndo,
+  liftRedeemerKleisli,
+  liftRedeemerCheck,
+  liftScriptContextEndo,
+  liftScriptContextKleisli,
+  liftScriptContextCheck,
+  liftValidatorEndo,
+  liftValidatorKleisli,
+  liftValidatorCheck,
+  liftMintingPolicyEndo,
+  liftMintingPolicyKleisli,
+  liftMintingPolicyCheck,
+  -- ** Specific checkers
+  txInfoRedeemersAreSorted,
+  txInfoOutputsAreSorted,
+  -- ** Specific normalizers
+  sortTxInfoRedeemers,
+  sortTxInfoOutputs,
+  -- * Elimination
+  runValidatorTransformer,
+  runValidatorTransformerPure,
+  runMintingPolicyTransformer,
+  runMintingPolicyTransformerPure
+  ) where
+
+import Control.Applicative ((<|>))
+import Data.Void (Void, absurd)
+import Control.Monad.Except (MonadError (throwError))
+import Data.Bifunctor (first)
+import Control.Arrow (Kleisli (Kleisli))
+import Data.Kind (Type)
+import Control.Category ((>>>), id)
+import Prelude hiding (id)
+import Data.Profunctor (Star (Star))
+import PlutusLedgerApi.V2 (ScriptContext (scriptContextTxInfo), 
+  txInfoRedeemers, ScriptPurpose (Minting, Spending, Rewarding, Certifying), 
+  txInfoOutputs, txOutReferenceScript)
+import Data.Profunctor.Rep (tabulate)
+import Data.Monoid (Endo (Endo, appEndo))
+import PlutusTx.AssocMap qualified as AssocMap
+import Data.List (sort, sortOn)
+
+-- | @since 3.0.1
+newtype Transformer (d :: Type) (r :: Type) (err :: Type) = 
+  Transformer (Star (Either err) (Maybe d, r, ScriptContext) (Maybe d, r, ScriptContext))
+
+-- | Verifies that 'txInfoRedeemers' in a 'ScriptContext' are sorted.
+--
+-- @since 3.0.1
+txInfoRedeemersAreSorted :: forall (d :: Type) (r :: Type) (err :: Type) . 
+  -- | Error to produce on failure
+  err -> 
+  Transformer d r err
+txInfoRedeemersAreSorted onFailure = liftScriptContextCheck . Kleisli $ \sc -> do
+  let tirs = AssocMap.toList . txInfoRedeemers . scriptContextTxInfo $ sc
+  if isSorted tirs
+    then Nothing
+    else Just onFailure
+
+-- | Verifies that 'txInfoOutputs' are sorted on reference scripts.
+--
+-- @since 3.0.1
+txInfoOutputsAreSorted :: forall (d :: Type) (r :: Type) (err :: Type) . 
+  -- | Error to produce on failure
+  err -> 
+  Transformer d r err
+txInfoOutputsAreSorted onFailure = liftScriptContextCheck . Kleisli $ \sc -> do
+  let tios = txInfoOutputs . scriptContextTxInfo $ sc
+  if isSortedBy txOutReferenceScript tios
+    then Nothing
+    else Just onFailure
```

Comment:
Outputs don't get sorted, only inputs.

This actually highlights two things:
- we should probably have a bunch of known-good script contexts from SC tools and ensure that out checkers all pass on them and that the normalizers are no-ops
- We need a list of what these checks/normalizeres should be